aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -744,6 +746,72 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
 }
 
 DEFINE_DISPATCH(sparse_mask_intersection_out_stub);
+DEFINE_DISPATCH(sparse_mask_projection_out_stub);
+
+using OptTensor = c10::optional<Tensor>;
+
+std::tuple<Tensor, Tensor, OptTensor> sparse_mask_like_prepare_sparse_inputs(
+    const std::string& method_name,
+    const Tensor& t,
+    const Tensor& mask) {
+  // This is a helper function for operations that implement "sparse_mask"-like
+  // functionality, namely, projection of values of one tensor onto the other.
+  // These operations mostly rely on COO intersection primitives that heavily
+  // exploit coalesced inputs to avoid any syncs and calls to sort. The problem
+  // is that these primitives might project first argument onto second one or
+  // the other way around depending on which arguments are coalesced and which are
+  // larger. This function prepares inputs for `sparse_mask` such that `t` is
+  // projected onto `mask` by sorting `t` if uncoalesced and artifically marking it
+  // as coalesced all while `mask` is set to uncoalesced.
```

Comment:
What is the advantage of returning `mask` as uncoalesced?

Reply:
If `mask` is uncoalesced, but the other argument is not, the COO intersection kernel will return a tensor with the same indices as `mask` but will do binary search of `mask` hashes into the hashes of the other arguments all without calls to `sort` and `coalesce`. The COO intersection kernel is heavily optimized to take advantage of `is_coalesced` of either arguments.

Reply:
This (forcing a coalesced input to be uncoalesced) sounds like a convoluted way to control COO intersection kernel functionality (to ensure that the result indices are the input indices).

Reply:
It still better than writing things from scratch, imho. The COO intersection kernel is not a public function with fixed semantics.